From dd1c45b6b5c4f5b71f3922af728fdcd73a263aa9 Mon Sep 17 00:00:00 2001
From: Brian Waldon <brian.waldon@rackspace.com>
Date: Tue, 25 Oct 2011 22:41:42 -0400
Subject: [PATCH] Remove 'location' from POST/PUT image responses

The 'location' field is already removed from GET calls, so we should
also remove it from POST/PUT operations. Partially fixes bug 880910.

(cherry picked from commit 258aa1356ca0e3e6de0b1cdd54b3736e592d3995)

Change-Id: I4f7d8d0309c8a3e10d0c2a99573ca0fa808c93be
---
 glance/api/v1/images.py           |    8 ++++++++
 glance/tests/unit/test_api.py     |   29 +++++++++++++++++++++++------
 glance/tests/unit/test_clients.py |    4 +---
 3 files changed, 32 insertions(+), 9 deletions(-)

diff --git a/glance/api/v1/images.py b/glance/api/v1/images.py
index 5d177e1..34efd5f 100644
--- a/glance/api/v1/images.py
+++ b/glance/api/v1/images.py
@@ -540,6 +540,10 @@ class Controller(api.BaseController):
             if location:
                 image_meta = self._activate(req, image_id, location)
 
+        # Prevent client from learning the location, as it
+        # could contain security credentials
+        image_meta.pop('location', None)
+
         return {'image_meta': image_meta}
 
     def update(self, req, id, image_meta, image_data):
@@ -585,6 +589,10 @@ class Controller(api.BaseController):
         else:
             self.notifier.info('image.update', image_meta)
 
+        # Prevent client from learning the location, as it
+        # could contain security credentials
+        image_meta.pop('location', None)
+
         return {'image_meta': image_meta}
 
     def delete(self, req, id):
diff --git a/glance/tests/unit/test_api.py b/glance/tests/unit/test_api.py
index 5da679d..94bcd59 100644
--- a/glance/tests/unit/test_api.py
+++ b/glance/tests/unit/test_api.py
@@ -1975,10 +1975,6 @@ class TestGlanceAPI(unittest.TestCase):
         res = req.get_response(self.api)
         self.assertEquals(res.status_int, httplib.CREATED)
 
-        res_body = json.loads(res.body)['image']
-        self.assertEquals(res_body['location'],
-                          'file:///tmp/glance-tests/3')
-
         # Test that the Location: header is set to the URI to
         # edit the newly-created image, as required by APP.
         # See LP Bug #719825
@@ -2056,6 +2052,29 @@ class TestGlanceAPI(unittest.TestCase):
         self.assertEqual(res.status_int, 200)
         self.assertFalse('X-Image-Meta-Location' in res.headers)
 
+        # Check PUT
+        req = webob.Request.blank("/images/2")
+        req.body = res.body
+        req.method = 'PUT'
+        res = req.get_response(self.api)
+        self.assertEqual(res.status_int, 200)
+        res_body = json.loads(res.body)
+        self.assertFalse('location' in res_body['image'])
+
+        # Check POST
+        req = webob.Request.blank("/images")
+        headers = {'x-image-meta-location': 'http://localhost',
+                   'x-image-meta-disk-format': 'vhd',
+                   'x-image-meta-container-format': 'ovf',
+                   'x-image-meta-name': 'fake image #3'}
+        for k, v in headers.iteritems():
+            req.headers[k] = v
+        req.method = 'POST'
+        res = req.get_response(self.api)
+        self.assertEqual(res.status_int, 201)
+        res_body = json.loads(res.body)
+        self.assertFalse('location' in res_body['image'])
+
     def test_image_is_checksummed(self):
         """Test that the image contents are checksummed properly"""
         fixture_headers = {'x-image-meta-store': 'file',
@@ -2076,8 +2095,6 @@ class TestGlanceAPI(unittest.TestCase):
         self.assertEquals(res.status_int, httplib.CREATED)
 
         res_body = json.loads(res.body)['image']
-        self.assertEquals(res_body['location'],
-                          'file:///tmp/glance-tests/3')
         self.assertEquals(image_checksum, res_body['checksum'],
                           "Mismatched checksum. Expected %s, got %s" %
                           (image_checksum, res_body['checksum']))
diff --git a/glance/tests/unit/test_clients.py b/glance/tests/unit/test_clients.py
index dd2e871..a7c4b1b 100644
--- a/glance/tests/unit/test_clients.py
+++ b/glance/tests/unit/test_clients.py
@@ -560,7 +560,6 @@ class TestRegistryClient(unittest.TestCase):
                    'container_format': 'ovf',
                    'status': 'active',
                    'size': 19,
-                   'location': "file:///tmp/glance-tests/2",
                    'properties': {}}
 
         images = self.client.get_images_detailed()
@@ -787,7 +786,6 @@ class TestRegistryClient(unittest.TestCase):
                    'container_format': 'ami',
                    'status': 'active',
                    'size': 13,
-                   'location': "swift://user:passwd@acct/container/obj.tar.0",
                    'properties': {'type': 'kernel'}}
 
         data = self.client.get_image(1)
@@ -811,7 +809,6 @@ class TestRegistryClient(unittest.TestCase):
                    'disk_format': 'vmdk',
                    'container_format': 'ovf',
                    'size': 19,
-                   'location': "file:///tmp/glance-tests/acct/3.gz.0",
                   }
 
         new_image = self.client.add_image(fixture)
@@ -844,6 +841,7 @@ class TestRegistryClient(unittest.TestCase):
         # Test ID auto-assigned properly
         self.assertEquals(3, new_image['id'])
 
+        del fixture['location']
         for k, v in fixture.items():
             self.assertEquals(v, new_image[k])
 
-- 
1.7.6.5

